dgelreset/dgelreset.1
.\" @(#)dgelreset.1	1.2 4/23/01 Copyright (C) 1997-2001 Digi International Inc., All Rights Reserved;
.TH DGELRESET 1 "30 October 2001"
.ds ]W Digi International
.SH NAME
dgelreset \- initiates a unit reboot in Digi EtherLite(R) modules
.SH SYNOPSIS
.B dgelreset
[
.I -port
tcp_port ]
.I ip_addr
.SH DESCRIPTION
The
.B dgelreset
utility will cause the named unit to reboot.  The functionality is
generally used to cause an EtherLite(R) module to re-request its
firmware via TFTP (which occurs during the module startup process).

The address parameter (
.I ip_addr
) may be supplied
in Internet standard (decimal) dot notation (e.g., 192.9.200.123),
or as an IP name.

.SH "OPTIONS"

.TP 17
.IR -port\  "tcp_port"
to specify the RealPort TCP port if the value has been changed from
its default (771) in the remote unit.

.SH "ADDITIONAL NOTES"

.B dgelreset
may hang if the module is already awaiting an IP address.

.SH "PROVIDED BY"
.sp
Technical Support
.br
Digi International Inc.
.br
11001 Bren Road E.
.br
Minnetonka, MN 55343
.br
Phone: 1-952-912-3456
.br
Fax: 1-952-912-4958
.br
email: support@digi.com
.br
WWW: http://www.digi.com/
.br
FTP: ftp.digi.com
.sp


release_notes.txt
                      Release Notes PN 93000359_X6P

                 Digi RealPort Driver Package for Linux
                            dgrp-1.9.1-36.i386.rpm

                       Tested Linux Distributions:
                 Red Hat Enterprise Linux 2.1, 3, 4, 5.x, 6.x, 7.x
                      Red Hat Linux 7.2, 7.3, 8, 9
                 Fedora Core 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 17
       Mandriva Linux 8.2, 9.0, 9.1, 9.2, 10.0, 10.1, 2006, 2007.1, 2008
            SuSE 8.1, 8.2, 9.0, 9.1, 9.2, 9.3, 10.0, 10.1, 10.2, 11.1, 11.2
                           Debian 3.0.r1, 3.1, 4, 6.0
                        Ubuntu 7.04, 7.10, 8.04, 8.10, 12.04. 18.04


                         Linux Kernels supported:
                            2.4.x (UP and SMP)
                            2.6.x (UP and SMP)
                            3.x.x (UP and SMP)
                            4.x.x (UP and SMP)

                       RPM Part Number 40002086_X6P

                              01/24/2019


  CONTENTS
  
  Section    Description
  1          Introduction
  2          Supported Products
  3          Enhancements
  4          Bug Fixes
  5          Known Limitations
  6          Additional Information
  7          History



  1. INTRODUCTION    

     Digi RealPort Linux is a driver package for Digi's Ethernet-based
     Serial Products.
     
     It is assumed that TCP/IP is running on the system and is properly
     configured.
     
     It is also useful, though not necessary, for the Wish package to
     be installed, or some other TCL interpreter, in order to use the
     RealPort Manager configuration tool.
     
     It is currently supported on the following hardware platforms:
     
     o Standard i386/i486 and Pentium PC (x86 32bit)
     o x86 64bit
     
     and is currently supported on the following Linux distributions:
     
     o Red Hat Enterprise Linux 2.1, 3, 4, 5.x, 6.x, 7.x
     o Red Hat Linux 7.2, 7.3, 8, 9
     o Fedora Core 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 17, 18
     o Mandriva Linux 8.2, 9.0, 9.1, 9.2, 10.0, 10.1, 2006, 2007.1, 2008
     o SuSE 8.1, 8.2, 9.0, 9.1, 9.2, 9.3, 10.0, 10.1, 10.2, 11.1, 11.2
     o Debian 3.0.r1, 3.1, 4, 6
     o Ubuntu 7.04, 7.10, 8.04, 8.10, 12.04, 18.04
     
     NOTE: Because of the rapid rate of releases from each respective
           Vendor, the tested/supported list above quickly becomes
           out of date.
           
           This driver package has been tested and verified working for
           kernels up to and including version 2.4.37, 2.6.38, and on
           various 3.X kernels (3.0.0, 3.0.58 and 3.8.0, 3,9.0, 3.10.0,
           3.11.0, 3.12.0, 3.13.0 and 3.14-rc4) from kernel.org.
           
           It is anticipated that this driver will work with Vendor kernel
           releases newer than the above but cannot be guaranteed because
           each respective Vendor can, and does, add their own various
           changes/patches to the stock kernel.org kernel.
           
           These additional Vendor patches to the stock kernel.org linux
           kernel can cause unforeseen incompatibilities with this driver.
     
     Please reference the following number(s) when searching the Digi
     International web site (www.digi.com) or ftp site (ftp.digi.com)
     for the latest software package:
     
     RPM Part Number: 40002086_X6P


  2. SUPPORTED PRODUCTS

     Digi Connect Family
     Digi EtherLite Family
     Digi Flex Family
     Digi One Family
     Digi CM Family
     Digi PortServer Family
     Digi PortServer II Family
     Digi PortServer TS Family
     Digi PortServer TS MEI Family
     Digi PortServer TS W MEI Family
     Digi PortServer TS M MEI Family
     Digi PortServer TS H MEI Family
     Digi PortServer TS P MEI Family
     IBM LAN Attached RAN 16


  3. ENHANCEMENTS
     o None

  4. BUG FIXES
     o Fix an installation problem on systems such as RHEL 7 where
       /etc/sysconfig/network exists but does not define a NETWORKING
       environment variable.

     o Fix a driver build problem on newer kernels, which removed
       definitions for TTY_CLOSING (in v3.11) and TTY_HW_COOK_IN (in v3.19).
       Note that Red Hat backported at least some of these definition
       removals into RHEL 7 kernels starting sometime shortly after
       3.10.0-229, so this driver fix is also needed for their newer
       3.10 kernels.

     o Add a description of the rtstoggle option to the ditty man page.

     o Fix a timing vulerability where an application that fast cycles
       many port open/closes could occasionally obstruct a close from
       completing properly, thus leading to a "locked" port.

     o Fix a problem that could cause the dpa.dgrp utility program to
       crash with an arithmetic exception if it happened to catch a port
       during the early stages of an uncompleted open.
      

  5. KNOWN LIMITATIONS

     o To build the dgrp driver you must have first loaded the matching
       kernel source tree for your target kernel.

     o To verify that you have matching kernel source, run "/bin/uname -r"
       and check to make sure the /usr/src directory has the corresponding
       linux-<version> directory.

     o If you are running a prebuilt kernel from a packaged distribution,
       this typically involves loading the kernel sources from your
       distribution CD to your system, then using the following commands
       to do the following:

       1. cd /usr/src/linux-<your_version_of_the_kernel>
       2. make mrproper           Clean up any old version files.
       3. make oldconfig          Make a configuration file to match
                                  your running kernel.
       4. make dep                Create the dependency and version files.

       Now you can load and build the dgrp driver package.


     o Occasionally, a Linux Vendor will ship a kernel that simply cannot

dgelreset/dgelreset.c
/*****************************************************************************
 *
 *  dgelreset - a utility program with one function:
 *    1. Reboot an Etherlites unit with either FAS or RealPort firmware
 *
 *  Authored by James Puzzo, October 2001, though much is borrowed
 *  directly from "cdipserv" and "cdelsreset" in the Central Data
 *  EtherLte software tool kit.
 * 
 * Copyright (c) 2001-2002 Digi International Inc., All Rights Reserved
 *
 *****************************************************************************/


#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>			/* for malloc() */
#include <string.h>
#include <errno.h>          /* for errno */
#include <time.h>           /* for time(), difftime() */
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/socket.h>		/* for socket() */
#include <netdb.h>			/* for gethostbyname() */
#include <sys/time.h>		/* for select() */
#include <netinet/in.h>
#include <arpa/inet.h>		/* for inet_ntoa() */
#include <arpa/tftp.h>		/* for struct tftphdr and various constants */
#include <sys/param.h>		/* for MAXHOSTNAMELEN on Irix */
#include "fas.h"

#if defined(AIX4)
#include <net/if.h>		/* for struct ifnet */
#include <sys/select.h>		/* for struct fd_set */
#endif


#define MAX_MSG_SIZE    (3*512) /* Maximum packet size */
#define MAX_EL_REQUEST_GAP  60  /* seconds */

#define RP_TCP_PORT     771
#define ELS_TCP_PORT    10001


typedef enum { ALL, FAS, NETCX } RESET_TYPE;

RESET_TYPE reset_type = ALL;


/*-------------------------------------------------------------------*/
/* function prototypes */

int reset_elmodule   (struct hostent *ipinfo, int rp_port);
int reset_rp_elmodule(struct hostent *ipinfo, int rp_port);

void usage_and_exit(char **argv);


/*-------------------------------------------------------------------*/
int main(int argc, char *argv[]) 
{
  struct hostent *ipinfo;        /* IP info about remote host */
  int rp_port = RP_TCP_PORT;     /* the port for RealPort requests */

  char *str_ipaddr=NULL;         /* IP address from the command line */

  int j;                         /* general-purpose integer :-) */

  /*
   * Try to grok the command-line arguments.
   */
  for (j=1; j<argc; j++) {
    if (j==(argc-1)) {str_ipaddr=argv[j]; continue;}
    
    /*
     * -port 
     */
    if (strncmp("-p", argv[j], 2)==0) {
      j++;
      rp_port=(int)strtol(argv[j],NULL,0);
      continue;
    }

    /*
     * -rp 
     */
    if (strncmp("-r", argv[j], 2)==0) {
      if (reset_type != ALL && reset_type != NETCX) {
        printf("Only one of the following options may be used: -rp, -sts\n\n");
        usage_and_exit(argv);
      }
      reset_type = NETCX;
      continue;
    }

    /*
     * -sts 
     */
    if (strncmp("-s", argv[j], 2)==0) {
      if (reset_type != ALL && reset_type != FAS) {
        printf("Only one of the following options may be used: -rp, -sts\n\n");
        usage_and_exit(argv);
      }
      reset_type = FAS;
      continue;
    }

    printf("Unrecognized argument: \"%s\"\n\n", argv[j]);
    usage_and_exit(argv);
  } /* looping through argv[] */
 
 
  /*
   * Got enough args?
   */
  if (str_ipaddr==NULL) {
    printf("Too few arguments.\n\n");
    usage_and_exit(argv);
  }
  
  
  /*
   * Make sure the IP address is kosher.
   */
  ipinfo=gethostbyname(str_ipaddr);
  if (ipinfo==NULL) {
    printf("Bad IP name or address: \"%s\"\n", str_ipaddr);
    herror(argv[0]);
    usage_and_exit(argv);
  }
 
 
  printf("Resetting the EtherLite module...\n");
  if (reset_elmodule(ipinfo,rp_port)<0)
    printf("Remote reset failed.  You will have to manually power cycle "
           "the module.\n");
  
  return(0);  /* That's it! */
}


/*-------------------------------------------------------------------*/
#if defined(AIX4) || defined(HPUX) || defined(SCO) || defined(UW) || defined(Linux) || defined(TRU64)
# include <sys/ioctl.h>
#else
# include <sys/sockio.h>
#endif
#include <net/if_arp.h>
#define bcopy(s,d,l)    memcpy(d,s,l)
#define bzero(p,l)      memset(p,0,l)


/*---------------------------------------------------------------*/

int sock_read(int s, caddr_t buf, int wanted);
void bitmash(unsigned char *challenge, unsigned char *key);


int reset_elmodule(struct hostent *ipinfo, int rp_port) 
{
  int s;
  struct sockaddr_in el_addr;
  req_unit_unlock	unlock_req;
  rsp_unit_unlock	unlock_rsp;
  req_unit_inquiry	inquiry_req;
  rsp_unit_inquiry	inquiry_rsp;
  req_global		reset_req;
  rsp_global		reset_rsp;
  char              str_inquiry[64];

  int rc;


  if (reset_type != ALL && reset_type != FAS) {
    printf("Skipping attempt to use STS software in the module...\n");
    rc = reset_rp_elmodule(ipinfo,rp_port);
    return(rc);
  }
  
  /*
   * Create the socket and the IP connect to the module.
   */
  if ((s=socket(AF_INET, SOCK_STREAM, 0))<0) {
    perror("socket");
    return(-1);
  }
  
  memset((char *)&el_addr, 0, sizeof(el_addr));
  memcpy((char *)&el_addr.sin_addr, ipinfo->h_addr, ipinfo->h_length);

  el_addr.sin_family = AF_INET;
  el_addr.sin_port = htons(ELS_TCP_PORT);
  
  if (connect(s, (struct sockaddr *)&el_addr, sizeof(el_addr)) < 0) {
    close(s);

    /*
     *  If we can't connect to the FAS port, it might be a RealPort
     *  based unit.
     */
    rc = reset_rp_elmodule(ipinfo,rp_port);
    return(rc);
  }
  
  /*
   * Request a challenge from the ELS.
   */
  
  unlock_req.opcode = FAS_UNIT_UNLOCK;
  if (write(s, &unlock_req, FASCMDSZ) < 0) {
    perror("write");
    close(s);
    return(-1);
  }
  
  if (sock_read(s, (caddr_t)&unlock_rsp, FASCMDSZ) < 0) {
    perror("read");
    close(s);
    return(-1);
  }
  
  if (unlock_rsp.opcode != FAS_UNIT_UNLOCK) {
    printf("Bad unlock response received from EtherLite module.\n");
    close(s);
    return(-1);
  }
  
  /*
   * Do the math that'll prove to the module that we're kosher.
   */
  bitmash(unlock_rsp.challenge,inquiry_req.key);
  
  /*
   * Send back our response.
   */
  inquiry_req.opcode = FAS_UNIT_INQUIRY;
  if (write(s, &inquiry_req, FASCMDSZ) < 0) {
    perror("write");
    close(s);
    return (-1);
  }
  
  if (sock_read(s, (caddr_t)&inquiry_rsp, FASCMDSZ) < 0) {
    perror("read");
    close(s);
    return(-1);
  }
  
  if (inquiry_rsp.opcode != FAS_UNIT_INQUIRY) {
    printf("Bad inquiry response received from EtherLite module.\n");
    close(s);
    return(-1);
  }
  
  /* Consume the remainder of the FAS_UNIT_INQUIRY response. */
  if (sock_read(s, (caddr_t)str_inquiry,
		256*inquiry_rsp.cnt_hi+inquiry_rsp.cnt_lo) < 0) {
    perror("read");
    close(s);
    return(-1);
  }
  
  /*
   * Done with the challenge.  Now we can issue the reset command.
   */
  reset_req.opcode = FAS_UNIT_RESET;
  if (write(s, &reset_req, FASCMDSZ) < 0) {
    perror("write");
    close(s);
    return(-1);
  }
  
  bzero((char *)&reset_rsp, sizeof(rsp_global));
  if (sock_read(s, (caddr_t)&reset_rsp, FASCMDSZ) < 0) {
    perror("read");
    close(s);
    return(-1);
  }
  
  if (reset_rsp.opcode != FAS_UNIT_RESET) {
    printf("Bad reset response received from EtherLite module.\n"
	   "You will need to upgrade the firmware in the module to version 6.0\n"
	   "or later before you can reset it remotely.\n");
    close(s);
    return(-1);
  }

  printf("Reset complete.\n");  
  close(s);
  return(0);
}

int reset_rp_elmodule(struct hostent *ipinfo, int rp_port) 
{
  int s;
  struct sockaddr_in el_addr;
  unsigned char pktbuf[1024];
 
#if defined(DEBUG) 
  int rlen;
  int i;
  char nums[80];
  char chrs[80];
#endif

  if (reset_type != ALL && reset_type != NETCX) {
    printf("Skipping attempt to use RealPort software in the module...\n");
    return(-1);
  }

  printf("Communicating with the RealPort software in the module...\n");

  /*
   * Create the socket and the IP connect to the module.
   */
  if ((s=socket(AF_INET, SOCK_STREAM, 0))<0) {
    perror("socket");
    return(-1);
  }

  memset((char *)&el_addr, 0, sizeof(el_addr));
  memcpy((char *)&el_addr.sin_addr, ipinfo->h_addr, ipinfo->h_length);

  el_addr.sin_family = AF_INET;
  el_addr.sin_port = htons(rp_port);
  
  if (connect(s, (struct sockaddr *)&el_addr, sizeof(el_addr)) < 0) {
    perror("connect");
    close(s);
    return(-1);
  }


  /*
   *  Build the packet which executes a unit reboot in the NetC/X protocol
   */
  pktbuf[0] = 0xfd;   /* KME packet   */
  pktbuf[1] = 0x00;   /* Module ID    */
  pktbuf[2] = 0x04;   /* Request type */
  pktbuf[3] = 0x00;   /* Board number */
  pktbuf[4] = 0x00;   /* Conc number  */
  pktbuf[5] = 0x00;   /* Unused       */
  pktbuf[6] = 0x00;   /* Addr 0       */
  pktbuf[7] = 0x00;   /* Addr 1       */
  pktbuf[8] = 0x00;   /* Addr 2       */
  pktbuf[9] = 0x00;   /* Addr 3       */
  pktbuf[10]= 0x00;   /* Size, MSB 0  */
  pktbuf[11]= 0x06;   /* Size, LSB 6  */
  pktbuf[12]= 'N';
  pktbuf[13]= 'O';
  pktbuf[14]= 'T';
  pktbuf[15]= 'F';
  pktbuf[16]= 'A';
  pktbuf[17]= 'S';

  if (write(s, pktbuf, 18) < 0) {
    perror("write");
    close(s);
    return(-1);
  }

  bzero(pktbuf, 1024);
  if (sock_read(s, pktbuf, 12) < 0) {
    perror("read");
    close(s);
    return(-1);
  }

  if ( pktbuf[0] != 0xfd ||
       pktbuf[2] != 0x04 ) {
    printf("Bad reset response received from EtherLite module.\n");
    close(s);
    return(-1);
  }

  if ( pktbuf[10]!= 0x00 ||
       pktbuf[11]!= 0x00 ) {
    printf("Bad reset response length %d received from EtherLite module.\n",
           pktbuf[10] * 256 + pktbuf[11]);
    close(s);
    return(-1);
  }

#if defined(DEBUG)
  rlen=12;
  nums[0]=chrs[0]=0;
  for (i=0;i<rlen;i++)
  { char dm[4];
    if( i && !(i % 16) )
    {
      printf("%-48.48s  %s\n", nums, chrs);
      nums[0]=chrs[0]=0;
    }

    sprintf(dm,"%02X ", (unsigned char)pktbuf[i]); strcat(nums, dm);

    if( isgraph( pktbuf[i] ) )
    {
      sprintf( dm, "%c", (unsigned char)pktbuf[i] );
      strcat(chrs,dm);
    }
    else strcat( chrs, "." );
  }

  if( i && (i % 16) )
  {
    printf("%-48.48s  %s\n", nums, chrs);
    nums[0]=chrs[0]=0;
  }
#endif /* DEBUG */

  printf("Reset complete.\n");
  close(s);
  return(0);
}

/*------------------------------------------------------------------*/
int sock_read(int s, caddr_t buf, int wanted) 
{
  int n;
  
  while (wanted)  {
    if ((n = read(s, buf, wanted)) <= 0) {
      return(-1);
    }
    buf += n;
    wanted -= n;
  }
  return(0);
}

/*------------------------------------------------------------------*/
/*
 *  bitmash function
 *
 *  Accepts an 6 byte challenge and returns a 3 byte key
 *  which has been bitmashed in a moderately inscrutable way.
 *  Certainly not enough security to foil a dedicated cryptographer
 *  but enough to discourage the casual cracker.
 *
 *  Those concerned with illicit access to the units should use the
 *  allowed hosts features implemented near the top of this file.
 *
 */
void bitmash(unsigned char *challenge, unsigned char *key) 
{
  int i, j, k, n, churns;
  
  /* mash the bits together */
  for( i=0, j=1, k=2, churns=0; churns < 31;
       i = (i + 1) % 6, j = (j + i) % 6, k = (k + j) % 6, churns++ ) {
    /* XOR a pair together */
    challenge[i] ^= challenge[j];
    /* rotate one byte each time */
    n = challenge[k] << 1;
    if( n & 0x100 )
      n |= 1;
    challenge[k] = (unsigned char) n;
  }
  /* build a key */
  key[0] = challenge[0] + challenge[4];
  key[1] = challenge[2] + challenge[1];
  key[2] = challenge[3] + challenge[5] + challenge[0];
  
  return;
}

/* give the user a clue & then exit */
void usage_and_exit(char **argv)
{
  printf("usage: %s [options] ip_addr\n", argv[0]);
  printf("  options can include:\n"
	 "    -port n   change the RealPort TCP port number to use\n"
    "              for EtherLite units with RealPort-based\n"
    "              firmware (default value is 771)\n"
    "    -rp       only attempt to reset by using the protocol\n"
    "              compatible with RealPort device drivers\n"
    "              (may not be used in conjunction with -sts)\n"
    "    -sts      only attempt to reset by using the protocol\n"
    "              compatible with STS device drivers\n"
    "              (may not be used in conjunction with -rp)\n"
	 "\n"	
	 "  ip_addr is the IP name or address of the unit to reset.\n"
	 "\n"	
	 );
  exit(1);
}
